clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatVersionedStats.java

```diff
@@ -23,4 +34,42 @@ public void recordLeaderLag(String storeName, int version, String region, long l
   public void recordFollowerLag(String storeName, int version, String region, long lag) {
     getStats(storeName, version).recordFollowerLag(region, lag);
   }
+
+  @Override
+  public synchronized void loadAllStats() {
+    Set<String> stores = getReportedStores();
+    metadataRepository.getAllStores().forEach(store -> {
+      if (stores.contains(store.getName())) {
+        addStore(store.getName());
+        updateStatsVersionInfo(store.getName(), store.getVersions(), store.getCurrentVersion());
+      }
+    });
+  }
+
+  @Override
+  public void handleStoreCreated(Store store) {
+    Set<String> stores = getReportedStores();
+    if (stores.contains(store.getName())) {
+      addStore(store.getName());
+    }
+  }
+
+  @Override
+  public void handleStoreChanged(Store store) {
+    Set<String> stores = getReportedStores();
+    if (stores.contains(store.getName())) {
+      updateStatsVersionInfo(store.getName(), store.getVersions(), store.getCurrentVersion());
+    }
+  }
+
+  Set<String> getReportedStores() {
+    if (leaderMonitors == null || followerMonitors == null) {
+      // TODO: We have to do this because theres a self call in the constructor
+      // of the superclass of this class. We shouldn't have to do this
+      return Collections.EMPTY_SET;
+    }
+    Set<String> monitoredStores = leaderMonitors.keySet();
+    monitoredStores.addAll(followerMonitors.keySet());
```

Comment:
`CHM::keySet` does not support additions, it is immutable, according to the JavaDoc and code... can you write a unit test that exercises `handleStoreChanged` (or any of the methods calling this function, besides the constructor) and see if this code actually works?

If the above bug suspicion is correct, then it can be fixed easily can copying the key set into a new set... but I feel like creating a combined set is for the sake of doing one `contains` on it is too expensive... can you refactor this so that we call `containsKey` on the two maps of interest instead?

Reply:
Sure

Reply:
Fixed